caddyhttp/proxy/reverseproxy.go

```diff
@@ -154,7 +155,12 @@ func NewSingleHostReverseProxy(target *url.URL, without string, keepalive int) *
 				prefer(target.RawPath, target.Path),
 				prefer(req.URL.RawPath, req.URL.Path))
 		}
-		req.URL.Path = singleJoiningSlash(target.Path, req.URL.Path)
+		req.URL.Path = singleJoiningSlash(target.Path,
+			prefer(req.Context().Value(staticfiles.URLPathCtxKey).(string), req.URL.Path))
+		// req.URL.Path must be consistent with decoded form of req.URL.RawPath if any
+		if req.URL.RawPath != "" && req.URL.RawPath != req.URL.EscapedPath() {
+			panic("RawPath doesn't match Path")
```

Comment:
BTW, i'm not sure but looks like that there are tests to test this case #1582, but as long as i understand they pass when they should not.
https://github.com/mholt/caddy/blob/master/caddyhttp/proxy/proxy_test.go#L518